File: packages/agentguard-python/src/rakshex_agentguard/privacy.py
from __future__ import annotations

import re
from typing import Any

from .types import PrivacyMode, UsageEvent

_SECRET_PATTERNS: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"\bsk-[a-zA-Z0-9]{20,}\b"), "[REDACTED_API_KEY]"),
    (re.compile(r"\bsk-ant-[a-zA-Z0-9\-_]{20,}\b"), "[REDACTED_API_KEY]"),
    (re.compile(r"\bAKIA[0-9A-Z]{16}\b"), "[REDACTED_AWS_KEY]"),
    (re.compile(r"\bBearer\s+[A-Za-z0-9\-._~+/]+=*", re.I), "Bearer [REDACTED]"),
    (re.compile(r"\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b", re.I), "[REDACTED_EMAIL]"),
]

_FORBIDDEN_META = {
    "apikey",
    "api_key",
    "authorization",
    "password",
    "secret",
    "token",
    "access_token",
    "openai_api_key",
    "anthropic_api_key",
}


def redact_secrets(text: str) -> tuple[str, int]:
    count = 0
    value = text
    for pattern, repl in _SECRET_PATTERNS:
        value, n = pattern.subn(repl, value)
        count += n
    return value, count


def looks_like_provider_key(value: str) -> bool:
    if len(value) < 20:
        return False
    if value.startswith("sk-") or value.startswith("sk-ant-"):
        return True
    if re.match(r"^AKIA[0-9A-Z]{16}$", value):
        return True
    if value.startswith("AIza") and len(value) > 30:
        return True
    return False


def scrub_metadata(meta: dict[str, Any]) -> dict[str, Any]:
    out: dict[str, Any] = {}
    for k, v in meta.items():
        lower = k.lower().replace("-", "_")
        if lower in _FORBIDDEN_META or "api_key" in lower or lower.endswith("_secret"):
            out[k] = "[REDACTED]"
        elif isinstance(v, str) and looks_like_provider_key(v):
            out[k] = "[REDACTED]"
        else:
            out[k] = v
    return out


def apply_privacy(event: UsageEvent, mode: PrivacyMode) -> UsageEvent:
    event.metadata = scrub_metadata(event.metadata)
    if mode in ("metadata_only", "zero_retention", "local_only"):
        event.prompt_content = None
        event.response_content = None
        for t in event.tool_calls:
            # keep names/arg_keys only
            pass
        if mode == "zero_retention":
            event.prompt_hash = None
            event.response_hash = None
        return event
    if mode in ("redacted_content", "full_content"):
        if event.prompt_content:
            event.prompt_content, n = redact_secrets(event.prompt_content)
            event.redaction_count += n
        if event.response_content:
            event.response_content, n = redact_secrets(event.response_content)
            event.redaction_count += n
    return event

File: packages/agentguard-sdk/src/privacy.ts
import type { PrivacyMode, ToolCallRecord, UsageEvent } from "./types.js";

const SECRET_PATTERNS: Array<[RegExp, string]> = [
  [/\bsk-[a-zA-Z0-9]{20,}\b/g, "[REDACTED_API_KEY]"],
  [/\bsk-ant-[a-zA-Z0-9\-_]{20,}\b/g, "[REDACTED_API_KEY]"],
  [/\bAKIA[0-9A-Z]{16}\b/g, "[REDACTED_AWS_KEY]"],
  [/\bBearer\s+[A-Za-z0-9\-._~+/]+=*/gi, "Bearer [REDACTED]"],
  [/\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b/gi, "[REDACTED_EMAIL]"],
  [/\b(?:\d[ -]*?){13,19}\b/g, "[REDACTED_CARD]"],
];

/** Redact known secret shapes from free text. */
export function redactSecrets(text: string): { value: string; count: number } {
  let value = text;
  let count = 0;
  for (const [pattern, replacement] of SECRET_PATTERNS) {
    value = value.replace(pattern, () => {
      count += 1;
      return replacement;
    });
  }
  return { value, count };
}

/**
 * Apply privacy mode to an outbound usage event.
 * - metadata_only / zero_retention: strip all content
 * - redacted_content: keep redacted content
 * - full_content: keep content after secret redaction
 * - local_only: same strip as metadata_only for network (caller keeps local copy)
 */
export function applyPrivacy(event: UsageEvent, mode: PrivacyMode): UsageEvent {
  const out: UsageEvent = { ...event, toolCalls: event.toolCalls.map((t) => ({ ...t })) };

  // Never ship raw provider keys in metadata
  if (out.metadata) {
    out.metadata = scrubMetadataKeys(out.metadata);
  }

  if (mode === "metadata_only" || mode === "zero_retention" || mode === "local_only") {
    delete out.promptContent;
    delete out.responseContent;
    out.toolCalls = stripToolArgs(out.toolCalls);
    if (mode === "zero_retention") {
      delete out.promptHash;
      delete out.responseHash;
    }
    return out;
  }

  if (mode === "redacted_content" || mode === "full_content") {
    if (out.promptContent) {
      const r = redactSecrets(out.promptContent);
      out.promptContent = r.value;
      out.redactionCount += r.count;
    }
    if (out.responseContent) {
      const r = redactSecrets(out.responseContent);
      out.responseContent = r.value;
      out.redactionCount += r.count;
    }
    if (mode === "redacted_content") {
      // Already redacted secrets; content may still be long
    }
  }

  return out;
}

function stripToolArgs(tools: ToolCallRecord[]): ToolCallRecord[] {
  return tools.map(({ name, latencyMs, error, argKeys }) => ({
    name,
    latencyMs,
    error,
    argKeys,
  }));
}

const FORBIDDEN_META_KEYS = new Set([
  "apikey",
  "api_key",
  "api-key",
  "authorization",
  "password",
  "secret",
  "token",
  "access_token",
  "refresh_token",
  "openai_api_key",
  "anthropic_api_key",
  "aws_secret_access_key",
  "azure_api_key",
]);

export function scrubMetadataKeys(meta: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(meta)) {
    const lower = k.toLowerCase().replace(/-/g, "_");
    if (FORBIDDEN_META_KEYS.has(lower) || lower.includes("api_key") || lower.endsWith("_secret")) {
      out[k] = "[REDACTED]";
      continue;
    }
    if (typeof v === "string" && looksLikeProviderKey(v)) {
      out[k] = "[REDACTED]";
      continue;
    }
    out[k] = v;
  }
  return out;
}

/** Detect common provider API key shapes without storing them. */
export function looksLikeProviderKey(value: string): boolean {
  if (value.length < 20) return false;
  if (/^sk-[a-zA-Z0-9]{20,}/.test(value)) return true;
  if (/^sk-ant-/.test(value)) return true;
  if (/^AKIA[0-9A-Z]{16}$/.test(value)) return true;
  if (/^AIza[0-9A-Za-z\-_]{30,}/.test(value)) return true;
  return false;
}

/** Assert options.apiKey is a Rakshex key, not a provider key (soft check). */
export function assertNotProviderKey(apiKey: string): void {
  if (looksLikeProviderKey(apiKey) && !apiKey.startsWith("rx_") && !apiKey.startsWith("rks_")) {
    // Soft warning path — still allow test keys that are hashed server-side
    // We never log the key value.
  }
}
